vulture_os/toolkit/api_parser/extrahop/extrahop.py

```diff
@@ -0,0 +1,205 @@
+
+import requests
+from requests.auth import HTTPBasicAuth
+
+logging.config.dictConfig(settings.LOG_SETTINGS)
+logger = logging.getLogger('api_parser')
+
+
+class ExtrahopAPIError(Exception):
+    pass
+
+
+class ExtrahopParser(ApiParser):
+
+    HEADERS = {
+        "Accept": "application/json"
+    }
+
+    def __init__(self, data):
+        super().__init__(data)
+
+        self.extrahop_host = data["extrahop_host"].rstrip("/")
+        if not self.extrahop_host.startswith('https://'):
+            self.extrahop_host = f"https://{self.extrahop_host}"
+
+        self.id = data['extrahop_id']
+        self.secret = data['extrahop_secret']
+
+        self.access_token = data.get('extrahop_access_token')
+        self.expire_date = data.get('extrahop_expire_date')
+
+        self.session = None
+
+    def get_access_token(self):
+        if not self.access_token or self.expire_date < timezone.now() + timedelta(minutes=5):
+            try:
+                url = f"{self.extrahop_host}/oauth2/token"
+                auth = HTTPBasicAuth(self.id, self.secret)
+                response = requests.post(url, auth=auth, data={"grant_type": "client_credentials"})
+                response.raise_for_status()
+            except requests.exceptions.HTTPError as err:
+                raise ExtrahopAPIError(f"Unable de renew access token. Error: {err}")
+            else:
+                response_json = response.json()
+                self.access_token = response_json["access_token"]
+                self.expire_date = (timezone.now() + timedelta(seconds=int(response_json["expires_in"]))).isoformat()
+                if self.frontend:
+                    self.frontend.access_token = self.access_token
+                    self.frontend.expire_date = self.expire_date
+                    self.frontend.save()
+
+        return self.access_token
+
+    def _execute_query(self, method, url, query, timeout=10):
+        """
+        raw request doesn't handle the pagination natively
+        """
+        max_retry = 3
+        retry = 1
+        while retry <= max_retry:
+            if self.session is None:
+                self._connect()
+
+            msg = f"call {url} with query {query} (retry: {retry})"
+            logger.debug(f"[{__parser__}]:__execute_query: {msg}", extra={'frontend': str(self.frontend)})
+            if method == "POST":
+                try:
+                    response = self.session.post(
+                        url,
+                        json=query,
+                        timeout=timeout,
+                        proxies=self.proxies,
+                        verify=self.api_parser_custom_certificate or self.api_parser_verify_ssl
+                    )
+
+                except Exception as e:
+                    logger.debug(f"[{__parser__}]:__execute_query: Exception: {e}", extra={'frontend': str(self.frontend)})
+                    self.session = None
+                    self.access_token = None
```

Comment:
don't treat **all** errors as a failure and invalidate token!
(also, you just invalidated the current runtime token and not the DB-stored one in `frontend`)